src/Mvc/Mvc.Core/src/HotReload/HotReloadService.cs

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Reflection.Metadata;
+using System.Threading;
+using Microsoft.AspNetCore.Mvc.Controllers;
+using Microsoft.AspNetCore.Mvc.Infrastructure;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+using Microsoft.AspNetCore.Mvc.ModelBinding.Metadata;
+using Microsoft.Extensions.Primitives;
+
+[assembly: MetadataUpdateHandler(typeof(Microsoft.AspNetCore.Mvc.HotReload.HotReloadService))]
+
+namespace Microsoft.AspNetCore.Mvc.HotReload
+{
+    internal sealed class HotReloadService : IActionDescriptorChangeProvider, IDisposable
+    {
+        private readonly DefaultModelMetadataProvider? _modelMetadataProvider;
+        private readonly DefaultControllerPropertyActivator? _controllerPropertyActivator;
+        private CancellationTokenSource _tokenSource = new();
+
+        public HotReloadService(
+            IModelMetadataProvider modelMetadataProvider,
+            IControllerPropertyActivator controllerPropertyActivator)
+        {
+            ClearCacheEvent += NotifyClearCache;
+            if (modelMetadataProvider.GetType() == typeof(DefaultModelMetadataProvider))
```

Comment:
This type isn't sealed?

Reply:
It's not. I've seen a couple users implement it. It feels weird to delete the caches under them.